### PR TITLE
fix: escape cells of CSV download

### LINKED ISSUE #69: Commas in names break the CSV export
**Describe the bug**
Verwendet einE Umfrage-TeilnehmerIn ein Komma im Namen, werden die Daten in dem CSV-Export nicht korrekt dargestellt. Das Komma wird als Ende der Zelle interpretiert, obwohl es übersprungen werden muss.

If a respondent uses a comma in their name, the data will not be displayed correctly in the CSV export. The comma is interpreted as the end of the cell, although it must be skipped.

**To Reproduce**

1. Create a poll
2. Answer the poll with the user name "te,st"
3. Download the csv and look for a broken dataset

**Expected behavior**
Commas should be escaped, so the user can use a comma at will. The csv should not break

**Desktop (please complete the following information):**
- all browsers

**Smartphone (please complete the following information):**
- all browsers

**Additional context**
All exported values like title, description and so on are affected.

### PR DESCRIPTION
**Pull request by jase88**
- escaping every CSV cell (untouched if not necessary)
- ensuring proper file name of CSV to download
  - for example too long file names that exceed 255 chars ([Long file names](https://en.wikipedia.org/wiki/Long_filename#:~:text=The%20maximum%20length%20of%20a,root%20folder%20of%20any%20drive.)) 
- made sure texts are localized 
- add unit test framework `jest` with Angulars experimental builder `@angular-devkit/build-angular:jest`
  - alternatively we could also 
    - a) use `jest-preset-angular` (which would add another package and some more configuration files) 
    - b) use `vitest` via `@analogjs/vite-plugin-angular` (which would add a layer of `analogjs` for testing)
  - happy to discuss this

fixes #69

### CODE REVIEW

**Comment by jase88** [comment]
@tjorbo ready for review or happy to discuss.
If necessary, I can also outsource things to separate PRs.

Bumping `rxjs` for example is also done here: https://github.com/Dataport/terminfinder-frontend/pull/200

File: src/locales/en-EN.json [hunk 1]
```diff
@@ -40,6 +40,16 @@
...
+    "csvExport": {
+      "fileTitlePrefix": "Poll",
+      "participants": "Participants",
+      "statusAccepted": "accepted",
+      "statusQuestionable": "questionable",
+      "statusDeclined": "declined",
+      "statusUnknown": "unknown",
+      "timeAt": "at {{time}}",
+      "dateTo": "to {{date}}"
+    },
```
Comment (tjorbo) [inline src/locales/en-EN.json:52]:
```suggestion
```

Please look at suggestions of de-DE-du.json - Thank you.

File: src/locales/platt.json [hunk 2]
```diff
@@ -39,6 +39,16 @@
...
+    "csvExport": {
+      "fileTitlePrefix": "Umfraag",
+      "participants": "Deelnehmer:innen",
+      "statusAccepted": ".",
+      "statusQuestionable": ".",
+      "statusDeclined": ".",
+      "statusUnknown": ".",
+      "timeAt": ".",
+      "dateTo": "."
+    },
```
Comment (tjorbo) [inline src/locales/platt.json:51]:
```suggestion
```

Please look at suggestions of de-DE-du.json - Thank you.

File: package.json [hunk 3]
```diff
@@ -80,8 +82,11 @@
     "express": "4.19.2",
     "font-awesome": "4.7.0",
     "http-server": "14.1.1",
+    "jest": "29.7.0",
```
Comment (tjorbo) [inline package.json:85]: It would be nice if you could rebuild the package-lock.json. Thank you.

**Review by tjorbo — CHANGES_REQUESTED** [review]
Hi @jase88,
Thank you for your PR.

I have reviewed your changes and agree with most of them.

Most of my requested changes are related to translations. Nice idea to translate the csv file as well. But is it possible to move the translation properties so we don't have to add a csvExport property?

Also, I found a serious bug while testing your changes. https://github.com/Dataport/terminfinder-frontend/issues/203 It is probably not a bug in your code, but if you could do a rebase, it would be clearer which changes are yours. We want to prevent merge commits by rebasing in general. So I will wait with rebasing dependabot's PRs until your PR is done. (To avoid conflicts in package.json)

Thanks again for your contribution.

File: src/app/poll/csv-utils.spec.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+describe("escapeCSVCell", () => {
+  it.each([
+    ["apple", "apple"],
+    ["banana", "banana"],
+    ["apple,banana", '"apple,banana"'],
+    ['orange"juice', '"orange""juice"'],
+    ["orange juice", "orange juice"],
+    ["line1\nline2", '"line1\nline2"'],
+    ["line1\rline2", '"line1\rline2"'],
```
[the remark this answers is not in the document]
Comment (jase88) [reply src/app/poll/csv-utils.spec.ts:11]: I agree that this implementation is very generic and mostly theoretical issues that may not occur in the project.

The API also seems to check that the name of a participant does not contain any control characters. The poll title does not seem to be checked in the API, but does not appear in the CSV export either

What do you suggest?  I prefer a general implementation that also covers changes to the features.

[the remark this answers is not in the document]
Comment (tjorbo) [reply src/app/poll/csv-utils.spec.ts:11]: You're right, I agree with your suggestion.

File: src/locales/platt.json [hunk 5]
```diff
@@ -93,6 +100,7 @@
   },
   "time": {
     "time": "Uhrtied",
+    "timeAt": ".",
```
[the remark this answers is not in the document]
Comment (tjorbo) [reply src/locales/platt.json:103]: It was an experimental feature that didn't make it to production. It's ok to be empty now.

File: src/app/poll/poll-form-helper.service.spec.ts [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    ];
+    service.appointment.title = "Title";
+    service.downloadCsv();
+
+    expect(createdDownloadAnchor.click).toHaveBeenCalled();
+    expect(createdDownloadAnchor.download).toBe("Umfrage-Title.csv");
+    expect(encodeURIComponentMock).toHaveBeenCalledWith(
+      [
+        'sep=,',
+        "Teilnehmende,01/10/2024 um 1:00 PM bis 01/11/2024 um 7:00 AM,01/20/2024 um 2:00 PM bis um 7:00 PM",
```
Comment (tjorbo) [inline src/app/poll/poll-form-helper.service.spec.ts:93]:
```suggestion
        `Teilnehmende,${moment("2024-01-10T12:00:00.000Z").format('MM/DD/YYYY [um] h:mm A')} bis ${moment("2024-01-11T06:00:00.000Z").format('MM/DD/YYYY [um] h:mm A')},${moment("2024-01-20T13:00:00.000Z").format('MM/DD/YYYY [um] h:mm A')} bis um ${moment("2024-01-21T18:00:00.000Z").format('h:mm A')}`,
```

GitHub Actions servers are on different time zones.

**Review by tjorbo — APPROVED** [review]

**Comment by jase88** [comment]
@tjorbo 
Also thank you for your patience and helping out with the mixed branch commits.

force pushing to my fork automatically closed this PR, wasn't aware of that.

So I pushed the branch you mentioned to my fork and created a new PR: https://github.com/Dataport/terminfinder-frontend/pull/209

Hope thats fine.

**Comment by tjorbo** [comment]
@jase88 I didn't know that either. Just merged #209. Thanks for the first contribution to this project. ❤️

**Pull request closed without merging**